mixins/provider-mixin.js

```diff
@@ -0,0 +1,38 @@
+export const ProviderMixin = superclass => class extends superclass {
+
+	constructor() {
+		super();
+		this._providers = new Map();
+		this.addEventListener('d2l-request-provider', this._handleRequest);
+	}
+
+	_handleRequest(e) {
+		if (this._providers.has(e.detail.key)) {
+			e.detail.provider = this._providers.get(e.detail.key);
+			e.stopPropagation();
+		}
+	}
+
+	provide(key, factory) {
+		this._providers.set(key, factory);
+	}
+
+	delete(key) {
```

Comment:
Since this will be "mixed in" to a lot of components, we want to be careful with the names of the methods that they won't collide. I'd recommend maybe `deleteProvider()` instead?

Reply:
And actually... what would the purpose of this be (other than testing)? The beauty of these providers being in the DOM and using standard DOM events is that once they're removed from the page, they stop providing.

Reply:
Hm, good question actually. @mdgbayly this doesn't look like it's used in the existing `attachments` case, so I presume it can just be removed?

Reply:
Yeh - i don't remember why i added in the delete method in the original provider I was playing with. I might have thought I needed it at one point, but then i don't call it anywhere anymore, so probably not required.